File: 1352-product-of-the-last-k-numbers/1352-product-of-the-last-k-numbers.py
class ProductOfNumbers:

    def __init__(self):
        self.cnt = [(1, 0)]  # (prefix product, zero count)

    def add(self, num: int) -> None:
        if num == 0:
            self.cnt.append((self.cnt[-1][0], self.cnt[-1][1] + 1))  # Keep same product, increment zero count
        else:
            self.cnt.append((self.cnt[-1][0] * num, self.cnt[-1][1]))  # Multiply product, keep zero count same

    def getProduct(self, k: int) -> int:
        if self.cnt[-1][1] - self.cnt[-(k + 1)][1] > 0:  # Check if zero exists in range
            return 0
        return self.cnt[-1][0] // self.cnt[-(k + 1)][0]  # Compute product using division


# Example Usage:
# obj = ProductOfNumbers()
# obj.add(3)
# obj.add(4)
# obj.add(2)
# obj.add(0)
# obj.add(5)
# obj.add(6)
# print(obj.getProduct(3))  # Output: 0
# print(obj.getProduct(2))  # Output: 30

File: 1352-product-of-the-last-k-numbers/1352-product-of-the-last-k-numbers.cpp
#include <vector>
using namespace std;

class ProductOfNumbers {
public:
    vector<unsigned long long> products;
    int zeroCount = 0;

    ProductOfNumbers() {
        products.push_back(1);
    }

    void add(int num) {
        if (num == 0) {
            products.clear();
            products.push_back(1);
            zeroCount = 0;
        } else {
            products.push_back(products.back() * num);
        }
    }

    int getProduct(int k) {
        if (k >= products.size()) return 0;
        return products.back() / products[products.size() - k - 1];
    }
};
